tests/component/specs/viewToolbar.cy.js

```diff
@@ -224,4 +224,93 @@ describe('View Toolbar Component', () => {
     // TODO: visual regression test
     // https://github.com/cylc/cylc-ui/issues/178
   })
+
+  it('filters task states', () => {
+    mountToolbar([
+      {
+        title: 'Group 1',
+        controls: [
+          {
+            title: 'Filter',
+            action: 'taskStateFilter',
+            value: [],
+            key: 'filter'
+          }
+        ]
+      },
+    ])
+    cy.get('.control > button').click()
+    cy.get('.v-list-item').as('menuItems')
+      .filter(':visible')
+      .should('have.length', 10)
+      .contains('.v-list-item', 'waiting').as('waitingItem')
+
+    // 1) select the waiting state and expand the waiting submenu
+    cy.get('@waitingItem')
+      .click()
+      .find('button')
+      .click({ force: true })
+
+      // the setOption event should be fired
+      .get('@wrapper').then(({ wrapper }) => {
+        expect(
+          wrapper.emitted().setOption.at(-1)[1]
+        ).to.deep.equal(['waiting'])
+      })
+
+      // the waiting state should be selected
+      .get('@waitingItem')
+      .should('have.class', 'v-list-item--active')
+
+      // the waiting submenu should be open
+      .get('.v-list-item').as('menuItems')
+      .filter(':visible')
+      .should('have.length', 15)
+
+    // 2) select the Retry state
+    cy.get('@menuItems')
+      .contains('Retry')
+      .click({ force: true })
+
+      // the setOption event should be fired
+      .get('@wrapper').then(({ wrapper }) => {
+        expect(
+          wrapper.emitted().setOption.at(-1)[1]
+        ).to.deep.equal(['waiting', 'isRetry'])
+      })
+
+    // 3) unselect the waiting state
+    cy.get('@waitingItem')
+      .click()
+
+      // implementation detail: the setOption event is fired again
+      // but the waiting state is not removed
+      .get('@wrapper').then(({ wrapper }) => {
+        expect(
+          wrapper.emitted().setOption.at(-1)[1]
+        ).to.deep.equal(['isRetry', 'waiting'])
+      })
+
+      // the waiting state should still be selected
```

Comment:
This doesn't seem right - surely the expected UX is that trying to unselect waiting should unselect it and its children?

Reply:
That's another way of doing it.

Happy either way.

Reply:
I think that's better if we have time, but this is better than nothing.

Reply:
In that case I think that this is the best we can get right now, and I'm happy to run with it.